waku/v2/protocol/waku_store/waku_store.nim

```diff
@@ -289,54 +289,80 @@ proc paginate*(list: seq[IndexedWakuMessage], pinfo: PagingInfo): (seq[IndexedWa
-      retrievedPageSize = min(uint64(pageSize), MaxPageSize).min(remainingMessages)  
-      if initQuery : foundIndex = foundIndex - 1
-      s = foundIndex + 1  # non inclusive
-      e = foundIndex + retrievedPageSize 
-      newCursor = msgList[e].index # the new cursor points to the end of the page
-    of PagingDirection.BACKWARD: # backward pagination
-      var remainingMessages = foundIndex 
-      if initQuery:  remainingMessages = remainingMessages + 1
-      # the number of queried messages cannot exceed the MaxPageSize and the total remaining messages i.e., foundIndex-0
-      retrievedPageSize = min(uint64(pageSize), MaxPageSize).min(remainingMessages) 
-      if initQuery : foundIndex = foundIndex + 1
-      s = foundIndex - retrievedPageSize 
-      e = foundIndex - 1
-      newCursor = msgList[s].index # the new cursor points to the begining of the page
-
-  if (retrievedPageSize == 0):
-    return (@[], PagingInfo(pageSize: 0, cursor:pinfo.cursor, direction: pinfo.direction), HistoryResponseError.NONE)
-
-  # retrieve the messages
-  var retMessages: seq[IndexedWakuMessage]= @[]
-  for i in s..e:
-    retMessages.add(msgList[i])
-  return (retMessages, PagingInfo(pageSize : retrievedPageSize, cursor : newCursor, direction : pinfo.direction), HistoryResponseError.NONE)
+      # set the index of the first message in the page
+      # exclude the message pointing by the cursor 
+      var startIndex = cursorIndex + 1
+      # for the initial query, include the message pointing by the cursor 
+      if isInitialQuery:  
+        startIndex = cursorIndex
+      
+      # adjust the pageSize based on the total remaining messages
+      pageSize = min(pageSize, total - startIndex)  
+
+      if (pageSize == 0):
+        return (@[], PagingInfo(pageSize: pageSize, cursor:pinfo.cursor, direction: pinfo.direction), HistoryResponseError.NONE)
+      
+      # set the index of the last message in the page
+      var endIndex = startIndex + pageSize - 1 
+
+      # retrieve the messages
+      var retMessages: seq[IndexedWakuMessage]
+      for i in startIndex..endIndex:
+        retMessages.add(msgList[i])
+      return (retMessages, PagingInfo(pageSize : pageSize, cursor : msgList[endIndex].index, direction : pinfo.direction), HistoryResponseError.NONE)
+
+    of PagingDirection.BACKWARD: 
+      # set the index of the last message in the page
+      # exclude the message pointing by the cursor 
+      var endIndex = cursorIndex - 1
+      # for the initial query, include the message pointing by the cursor
+      if isInitialQuery:  
+        endIndex = cursorIndex
+      
+      # adjust the pageSize based on the total remaining messages
+      pageSize = min(pageSize, endIndex + 1) 
+
+      if (pageSize == 0):
+        return (@[], PagingInfo(pageSize: pageSize, cursor:pinfo.cursor, direction: pinfo.direction), HistoryResponseError.NONE)
```

Comment:
Can we pull this up into an intermediate variable and return that? Makes debugging easier

Reply:
Of course, done!